brainstorming.md

```diff
@@ -0,0 +1,508 @@
+
+> ⚠️ **Issue** - the above uses Arrow `string`. Should we be using `large_string` instead (no 2GB cap)?
+
+## Metadata
+
+All SOMA objects may be annotated with a small amount of metadata:
+
+- `metadata` (map[string, simple]) - user metadata, specified as string-keyed mapping to any [**simple**](#type-definitions-used-in-this-document) type (eg, int32, string, etc).
+
+Only [`simple`](#type-definitions-used-in-this-document) types are supported as metadata values. The metadata lifecycle is the same as its containing object, eg, it will be deleted when the containing object is deleted.
+
+> ⚠️ **Issue** - do we need more complex metadata value types, and if so, how to represent capture what is supported? (eg, anything Arrow can represent? a subset?)
+>
+> **Proposal**: restrict to a mapping of string keys and `simple` values. Extend to complex types in the future as we have concrete use cases which warrant the complexity.
+
+## Foundational Types
+
+The foundational types represent the core data structures used to store and index data. They are intended to be moderately general purpose, and to serve as building blocks for the [Composed Types](#composed-types) which codify domain-specific use cases (eg, single cell experimental datasets).
+
+### SOMACollection
+
+SOMACollection is an unordered, `string`-keyed map of values. Values may be any SOMA foundational or composed type, including other (nested) SOMACollection objects. Keys in the map are unique and singular (no duplicates, ie, the SOMACollection is _not_ a multi-map). The SOMACollection is expected to be used for a variety of use cases:
+
+- as a container of independent objects (eg, a collection of single-cell datasets, each manifest as a [SOMAExperiment](#soma-experiment) object)
+- as the basis for building other composed types (eg, using SOMACollection to organize pre-defined fields in [SOMAExperiment](#soma-experiment) such as multiple layers of `X`).
+
+### SOMADataFrame
+
+`SOMADataFrame` is a multi-index, multi-column table of monomorphic arrays, all of equal length. A `SOMADataFrame` has a user-defined schema, which defines the number of columns and their respective column name and value type. The schema is expressed as an Arrow `Schema`. All `SOMADataFrame` contain a "pseudo-column" called `__rowid`, of type `uint64` and domain `[0, #rows)`. The `__rowid` pseudo-column contains a unique value for each row in the `SOMADataFrame`, and is intended to act as a join key for other objects, such as a `SOMANdArray`.
```

Comment:
did we define _monomorphic_ anywhere? since this is a doc about an API/standard, we should spell out what we mean to remove all doubt

Reply:
will add!